course_catalog/api.py

```diff
@@ -526,3 +530,192 @@ def parse_bootcamp_json_data(bootcamp_data, force_overwrite=False):
+    """
+    loaded_raw_jsons_for_course = []
+    last_modified_dates = []
+    uid = None
+    is_published = True
+    log.info("Syncing: %s ...", course_prefix)
+    # Collect last modified timestamps for all course files of the course
+    for obj in raw_data_bucket.objects.filter(Prefix=course_prefix):
+        # the "1.json" metadata file contains a course's uid
+        if obj.key == course_prefix + "0/1.json":
+            try:
+                first_json = safe_load_json(get_s3_object_and_read(obj), obj.key)
+                uid = first_json.get("_uid")
+                last_published_to_production = format_date(
+                    first_json.get("last_published_to_production", None)
+                )
+                last_unpublishing_date = format_date(
+                    first_json.get("last_unpublishing_date", None)
+                )
+                if last_published_to_production is None or (
+                    last_unpublishing_date
+                    and (last_unpublishing_date > last_published_to_production)
+                ):
+                    is_published = False
+            except:  # pylint: disable=bare-except
+                log.exception("Error encountered reading 1.json for %s", course_prefix)
+        # accessing last_modified from s3 object summary is fast (does not download file contents)
+        last_modified_dates.append(obj.last_modified)
+    if not uid:
+        # skip if we're unable to fetch course's uid
+        log.info("Skipping %s, no course_id", course_prefix)
+        return uid, is_published
+    # get the latest modified timestamp of any file in the course
+    last_modified = max(last_modified_dates)
+
+    # fetch JSON contents for each course file in memory (slow)
+    log.info("Loading JSON for %s...", course_prefix)
+    for obj in sorted(
+        raw_data_bucket.objects.filter(Prefix=course_prefix),
+        key=lambda x: int(x.key.split("/")[-1].split(".")[0]),
+    ):
+        loaded_raw_jsons_for_course.append(
+            safe_load_json(get_s3_object_and_read(obj), obj.key)
+        )
+
+    log.info("Parsing for %s...", course_prefix)
+    # pass course contents into parser
+    parser = OCWParser("", "", loaded_raw_jsons_for_course)
+    course_json = parser.get_master_json()
+    course_json["uid"] = uid
+    course_json["course_id"] = "{}.{}".format(
+        course_json.get("department_number"), course_json.get("master_course_number")
+    )
+
+    # if course run synced before, update existing Course instance
+    courserun_instance = CourseRun.objects.filter(course_run_id=uid).first()
+
+    # Make sure that the data we are syncing is newer than what we already have
+    if (
+        courserun_instance
+        and last_modified <= courserun_instance.last_modified
+        and not force_overwrite
+    ):
+        log.info("Already synced. No changes found for %s", course_prefix)
+        return uid, is_published
```

Comment:
If `uid` and `is_published` are returned here, the course will be either reindexed (or deindexed) even if no runs were changed.